### DISCUSSION TITLE
overrideIdFn with @formatjs/swc-plugin with storybook and Next

### DISCUSSION THREAD

**Discussion by williamhaley** [storybookjs/storybook#25772] [Help] [answered]
### Summary

Hello. It looks like `useSwc` (`swc` flag for storybook) is [enabled by default](https://storybook.js.org/docs/api/main-config-swc) for new webpack 5 storybook projects as of version 7.6. I'm opting-in to `useSwc` now to try things out.

```
...
  framework: {
    name: dirname(require.resolve(join('@storybook/nextjs', 'package.json'))),
    options: {
      builder: {
        useSWC: true
      }
    }
  },
...
```

However, I'm getting an error related to ids for `@formatjs/intl`.

```
[@formatjs/intl] An `id` must be provided to format a message. You can either:
1. Configure your build toolchain with [babel-plugin-formatjs](https://formatjs.io/docs/tooling/babel-plugin)
or [@formatjs/ts-transformer](https://formatjs.io/docs/tooling/ts-transformer) OR
2. Configure your `eslint` config to include [eslint-plugin-formatjs](https://formatjs.io/docs/tooling/linter#enforce-id)
to autofix this issue
```

I'm familiar with the issue at play here. Historically I use `babel` to handle the `id` for me so so that I do not need to explicitly specify an `id` for translation messages. Here's my current `babel` config in `main.ts` for storybook.

```
...
  babel: async options => ({
    ...options,
    plugins: [
      [
        'formatjs',
        {
          idInterpolationPattern: '[sha512:contenthash:base64:6]',
          ast: true
        }
      ]
    ]
  }),
...
```

With `useSwc` babel is ignored and so is my babel config. I need to adapt this for `swc`. [Storybook explains how to configure `swc` in the standard typescript `.storybook/main.ts`](https://storybook.js.org/docs/api/main-config-swc) file like so.

```
// Replace your-framework with the webpack-based framework you are using (e.g., react-webpack5)
import type { StorybookConfig } from '@storybook/your-framework';

const config: StorybookConfig = {
  framework: {
    name: '@storybook/your-framework',
    options: {
      builder: {
        useSWC: true, // This flag is automatically set by Storybook for all new Webpack5 projects (except Angular) in Storybook 7.6
      },
    },
  },
  swc: (config, options) => {
    return {
      ...config,
      // Apply your custom SWC configuration
    };
  },
};

export default config;
```

I think I'm failing to understand how to properly configure formatjs support for storybook. There is a [`@formatjs/swc-plugin` available](https://formatjs.io/docs/tooling/swc-plugin/), with a `Usage` example for using `overrideIdFn`. I'm not quite sure how/where to implement their example block within the storybook config though.

I've tried the following:

```
...
  swc: (config, options) => {
    return {
      ...config,
      jsc: {
        experimental: {
          plugins: [
            [
              '@formatjs/swc-plugin-experimental', {
                overrideIdFn: '[sha512:contenthash:base64:6]',
                ast: true
              }
            ]
          ]
        }
      }
    };
  },
...
```

Also of note is that `@formatjs/swc-plugin` specifically calls out a lack of Typescript support.

> @formatjs/swc-plugin is not currently compatible with TypeScript due to [swc-project/swc#4648](https://github.com/swc-project/swc/issues/4648). Attempting to compile a TypeScript codebase with the swc plugin will likely result in the error "Method visitTsType not implemented."

Despite that, I got the impression from searching that others may be having success in typescript so it was worth trying. I can't find where these were linked from, but I found a site pointing out [`@formatjs/swc-plugin-experimental`](https://www.npmjs.com/package/@formatjs/swc-plugin-experimental)  and [`@formatjs/swc-plugin`](https://www.npmjs.com/package/@formatjs/swc-plugin). Neither seems to change things for me though.

I tried to grok this issue, which seems relevant, https://github.com/formatjs/formatjs/issues/3589#issuecomment-1177485297, but didn't see any relevant answers.

[It seems like the syntax above _should_ work with storybook](https://swc.rs/docs/configuration/compilation#jscexperimentalplugins), but I haven't had any luck.

I feel like I'm missing something with the migration from babel to swc, but don't know what I need here. [Not sure if using Next.js adds another layer of complexity here](https://storybook.js.org/docs/configure/compilers#swc).

> Support for the SWC builder is currently experimental for Next.js projects, and it's not enabled by default. It requires you to opt in to use it. For more information on configuring SWC with the supported frameworks, see the [SWC API](https://storybook.js.org/docs/api/main-config-swc) documentation.

### Additional information

* `react-intl@6.1.1`
* `@formatjs/cli@4.8.2`
* `storybook@7.6.10` 
* `@storybook/nextjs@7.6.10`
* `next@14.0.3`

### Create a reproduction

Having trouble getting `storybook.new`'s Next + Typescript template working on Stackblitz. On Firefox and Chrome it keeps hanging on `Running start command` or `Booting WebContainer`. Not sure what's going on there, but I think I have a meaningful reproduction up here https://replit.com/@williamhaley1/Nextjs#.storybook/main.ts

##### useSwc: true

![2024-01-29_00-33-53_screenshot](https://github.com/storybookjs/storybook/assets/1851748/8105ffd8-f2d7-4a49-88b3-397c84321766)

#### useSwc: false

![2024-01-29_00-31-53_screenshot](https://github.com/storybookjs/storybook/assets/1851748/5c6cc774-834c-42c7-8d5d-dff22739977c)

**Comment by valentinpalkovic** [comment]
Can you tell me which Storybook version you have in use?

**Reply by williamhaley** [reply]
Updated the `Additional information`:

> * `react-intl@6.1.1`
> * `@formatjs/cli@4.8.2`
> * `storybook@7.6.10` 
> * `@storybook/nextjs@7.6.10`
> * `next@14.0.3`

**Reply by williamhaley** [reply]
I updated my post to add a reproduction that I think is meaningful. I seem to be able to stop the `storybook` instance in that reproduction, toggle `useSwc`, and see the issue appear/disappear

**Comment by valentinpalkovic** [comment]
Can you elaborate on how you configure `formatjs` for Next.js? Does Next.js just pick up your `.swcrc` configuration?

**Reply by valentinpalkovic** [reply]
Have you configured it in the `next.config.js` like described [here](https://nextjs.org/docs/architecture/nextjs-compiler#swc-plugins-experimental)? If so, I think Storybook should be able also to use the swcPlugins definition from your `next.config.js` to pass it further to the custom swc-compiler. Next.js is a bit special when we are talking about swc support. AFAIK, the `swc` preset in `.storybook/main.ts` doesn't apply for the Next.js framework.

We did some further Next.js/SWC optimizations in the latest Storybook 8.0 alpha. Would you like to try it out? Just run `npx storybook@next upgrade`.

**Reply by williamhaley** [reply, marked as answer]
> Can you elaborate on how you configure formatjs for Next.js? Does Next.js just pick up your .swcrc configuration?

I don't currently have a `.swcrc` configuration.

My actual use case (beyond the scope of a reproduction) is a [turborepo](https://turbo.build/repo/docs) monorepo. There are many `packages` and distinct Next.js `apps` in that one repo. Each of my dozen or so Next.js apps relies on a root-level `babel` config that handles formatjs id interpolation.

```
{
  "presets": ["next/babel"],
  "plugins": [
    [
      "formatjs",
      {
        "idInterpolationPattern": "[sha512:contenthash:base64:6]",
        "ast": true
      }
    ]
  ]
}
```

All my Next apps in the monorepo share the same base configuration, but they're all distinct apps with their own `next.config.js`.

I'm only now seeing babel is considered unsupported in the latest Next.js docs and swc is the preferred route. I can't remember if this was the case when I first adopted Next over a year ago.

> When your application has a .babelrc file, Next.js will automatically fall back to using Babel for transforming individual files. This ensures backwards compatibility with existing applications that leverage custom Babel plugins.
>
> https://nextjs.org/docs/architecture/nextjs-compiler#unsupported-features

But Next acknowledges many people use and rely on babel and its many plugins. I [see I'm not the only one who relies on this id interpolation behavior from babel with formatjs](https://github.com/vercel/next.js/discussions/30174#discussioncomment-1539820). I _can_ look into porting all of my many Next.js apps from babel -> swc, but that could be a huge impact that will need more testing.

I was happy to see I could configure `useSwc` in storybook to use `swc` specifically for my `storybook` package. It does seem to work as needed.

```
// main.ts

...

  framework: {
    name: getAbsolutePath("@storybook/nextjs"),
    options: {
      builder: {
        useSWC: true,
      },
    },
  },
```

```
...

> my-app@0.1.0 storybook
> storybook dev -p 3000

@storybook/cli v8.0.0-alpha.17

info => Starting manager..
info => Starting preview..
info Addon-docs: using MDX3
info => Using implicit CSS loaders
info => Using SWC as compiler
info => Using default Webpack5 setup

...
```

> We did some further Next.js/SWC optimizations in the latest Storybook 8.0 alpha. Would you like to try it out? Just run npx storybook@next upgrade.

I did try an upgrade to no avail. I'm on all the latest `storybook` plugins now, but no differences

> Have you configured it in the next.config.js like described [here](https://nextjs.org/docs/architecture/nextjs-compiler#swc-plugins-experimental)? If so, I think Storybook should be able also to use the swcPlugins definition from your next.config.js to pass it further to the custom swc-compiler. Next.js is a bit special when we are talking about swc support. AFAIK, the swc preset in .storybook/main.ts doesn't apply for the Next.js framework.

Thank you. This was most helpful. I actually wasn't configuring `next.config.js` at all for my `storybook` *package* within my monorepo. I was just letting storybook use whatever the implied defaults were. That single `storybook` package in my monorepo contains all my storybook stories and isn't a true Next.js app so it didn't have any `next.config.js` of its own.

I was trying to configure `swc` plugins in `main.ts` under the `swc` settings per https://storybook.js.org/docs/api/main-config-swc. But I do see that If I explicitly configure a `next.config.js` for my storybook package using a root-level `next.config.js`, then I *do* see that storybook uses it!

With that root-level config like :point_down:, storybook seems to use the formatjs swc wasm plugins! Unfortunately, I'm now convinced that I can't do what I want to do. At least, not yet.

```
// ./next.config.js

const nextConfig = {
  experimental: {
    swcPlugins: [
      [
        "@formatjs/swc-plugin-experimental",
        {
          overrideIdFn: "[sha512:contenthash:base64:6]",
        },
      ],
    ],
  },
};

module.exports = nextConfig;

```

This open issue https://github.com/formatjs/formatjs/issues/3589 and this discussion https://github.com/vercel/next.js/discussions/33698 are where I'm stuck. I'm realizing that `swc` is enabled for me in storybook, and the `@formatjs/swc-plugin-experimental` and the `@formatjs/swc-plugin` simply aren't working reliably yet. I'm seeing the same Rust stack traces as others in those issues/discussions.

**Sorry to waste your time. My original confusion was around how I might potentially configure swc specifically for storybook. That question is solved. I see that storybook's Next framework will load a root-level `next.config.js` file. Now that I know how I can configure the formatjs swc plugins for storybook the issues I have are dependent on formatjs updating their plugins and not on anything specific to storybook**
